Refactor `laundry.py` to use NamedTuple, add tests

tianyizheng02 opened the pull request:
Refactor `laundry.py` to store API output using `NamedTuple`s rather than plain Python `dict`s. This makes the output more user-friendly, less error-prone, and easier to type hint. This refactoring also modifies the API behavior as follows:
1. JSON objects that don't qualify as a known laundry machine type (washer, dryer, or combo) will now be skipped, whereas previously they would be marked as "unknown". These objects should be skipped because they may represent things that aren't laundry machines, such as card readers.
2. The `time_remaining` field for laundry machines can now be `None` if the laundry machine is unavailable (either out of service or offline).
3. The data for each machine now includes a `name` field that stores the user-facing name of the machine (A1, B3, G10, etc).

Add unit tests for `laundry.py` to increase test coverage to 95%. The new tests use mocked data for Towers laundry machines, the vast majority of which are combo machines rather than standalone washers or dryers.

Fix a heretofore unknown bug that was caught during the aforementioned unit testing: "washNdry" was misspelled as "washNDry", which was causing the API to mark all combo machines as "unknown". This typo hadn't been caught because there had been no unit tests that included data for combo machines.

tianyizheng02 committed:
Refactor laundry.py to use NamedTuple, add tests

Refactor laundry.py to store API output using NamedTuples rather than
plain Python dicts. This makes the output more user-friendly, less
error-prone, and easier to type hint. This refactoring also modifies the
API behavior as follows:
1. JSON objects that don't qualify as a known laundry machine type
   (washer, dryer, or combo) will now be skipped, whereas previously
   they would be marked as "unknown". These objects should be skipped
   because they may represent things that aren't laundry machines, such
   as card readers.
2. The time_remaining field for laundry machines can now be None if the
   laundry machine is unavailable (either out of service or offline).

Add unit tests for laundry.py to increase test coverage to 100%. The new
tests use mocked data for Towers laundry machines, the vast majority of
which are combo machines rather than standalone washers or dryers.

Fix a heretofore unknown bug that was caught during the aforementioned
unit testing: "washNdry" was misspelled as "washNDry", which was causing
the API to mark all combo machines as "unknown". This typo hadn't been
caught because there had been no unit tests that included data for combo
machines.

tianyizheng02 committed:
Fix indentation in laundry_test.py __init__

RitwikGupta commented on pittapi/laundry.py:
```diff
@@ -37,7 +37,22 @@
 
 
-def _get_laundry_info(building_name: str) -> Any:
+class BuildingStatus(NamedTuple):
```
Why is this class inheriting `typing.NamedTuple`? Did you mean `collections.NamedTuple`?
I think all you need/want is a `dataclass`. Also, how does this affect downstream serialization? That is, can I still dump this to a valid JSON object?

tianyizheng02 replied:
Yes, I meant to use `typing.NamedTuple` and not `collections.namedtuple`:
1. `typing.NamedTuple` has (imo) a cleaner syntax than `collections.namedtuple`.
2. `typing.NamedTuple` is typed, while `collections.namedtuple` isn't. One of my motivations for making this PR is to gradually prep the codebase so that we can introduce `mypy` to the repo for #45, and having explicitly defined types for data fields will help with that.
3. The PittAPI repo had already been using `typing.NamedTuple` long before I started contributing to it, so I'm also sticking to it for consistency.

`dataclass` would work for our purposes as well, and I don't think there's any effect on serialization in either case (both `dataclass` and `typing.NamedTuple` can be converted to plain Python `dict`, which could then be serialized to JSON). I don't know if `dataclass` provides any features/benefits that'd make it notably better than `typing.NamedTuple`, at least for our purposes.

RitwikGupta commented on pittapi/laundry.py:
```diff
@@ -37,7 +37,22 @@
+    total_dryers: int
+
+
+class LaundryMachine(NamedTuple):
```
Same as above, why is this class inheriting `typing.NamedTuple`?

RitwikGupta commented on pittapi/laundry.py:
```diff
@@ -61,39 +75,43 @@ def get_status_simple(building_name: str) -> dict[str, str]:
         # one part of combo being in use marks the whole thing as in use, so we can only show if
         # both parts are free.
+        #
+        # New: Above statement is not true, as the JSON objects provide two separate statuses for combo machines, one for the
```
Just get rid of the outdated comment, no need to say "new".

RitwikGupta reviewed (changes requested):
The TODOs need to be addressed before merging this PR. Mark this as WIP in the meantime.

tianyizheng02 committed:
Remove outdated comment

tianyizheng02 commented:
> The TODOs need to be addressed before merging this PR. Mark this as WIP in the meantime.

@RitwikGupta I've marked this PR as draft for the time being, but I can't really work on it any further anyway. The big problem here is that the TODOs are currently unaddressable, and I don't think they'll become addressable until the fall semester starts. I'll lay out my reasons here, mostly as a reminder to my future self:
- The first two TODOs pertain to updating the code to properly handle combo laundry machines. I know that the JSON laundry data provides separate JSON fields for each part of a combo machine, but I don't know which fields correspond to which part. A JSON object for a combo machine looks like this:
    ```json
    {
        "type": "washNdry",
        "appliance_type": "D",
        "model_number": "COMBO",
        "x": 173,
        "y": 35,
        "orientation": "NW",
        "appliance_desc_key": "10000010828",
        "appliance_desc": "GA03",
        "combo": true,
        "stacked": false,
        "opacity": 0,
        "status_toggle2": 0,
        "average_run_time2": 60,
        "time_remaining2": 60,
        "time_left_lite2": "Available",
        "percentage2": 0,
        "status_toggle": 0,
        "average_run_time": 60,
        "time_remaining": 60,
        "time_left_lite": "Available",
        "percentage": 0,
        "appliance_desc_key2": "10000010829",
        "appliance_desc2": "GA04"
    }
    ```
    From fields such as `time_remaining` and `time_remaining2`, we see that there are separate fields for the washer and the dryer—but I don't know which one is for the washer and which one is for the dryer. I can't think of any way to figure this out definitively that doesn't require someone going to campus and examining the combo machines in person.
- The third and final TODO is to figure out the meaning of a value of `1` for the `status_toggle` field. This case isn't covered in our mock test data, and I never encountered such a value when I tried querying the laundry API myself. I can't confirm the meaning definitively, but I _suspect_ that it means "Completed" (which would explain why I've never encountered it, since it's currently summer break and nobody is on campus doing laundry). Again, I don't think we can resolve this TODO until the fall semester starts, when people are actually using the laundry machines.

tianyizheng02 committed:
Add type hint to fix mypy --strict error

tianyizheng02 committed:
Complete TODOs for combo machines and status_toggle = 1

Replace mock JSON test data for Sutherland East with new mock JSON test
data for Holland. Because this data was collected close to the start of
the fall semester, when the laundry machines are actually being used, it
contains instances of status_toggle = 1, which clears up the TODO of
what that status value represented. The meaning of this value has now
been documented.

Rewrite the logic for parsing JSON objects for laundry machines. The
logic now accurately parses combo machines and adds support for double
machines, which are two washers or two dryers in a single machine. This
resolves the two remaining TODOs about combo machines.

tianyizheng02 commented:
@RitwikGupta I've addressed the TODOs. I was able to figure out the TODO for `status_toggle = 1` using laundry data from Holland that had the missing status value. I was able to figure out the TODOs for combo machines based on @jkruse7's observation in #135 about even-/odd-numbered machines.

RitwikGupta commented:
Excellent work @tianyizheng02 and @jkruse7!
